### PR TITLE
tip vote history population changes

### PR DESCRIPTION
**Pull request by sravanthikoneru**

### CODE REVIEW

File: HotTips/Engine/TipManager.cs [hunk 1]
```diff
@@ -24,7 +24,14 @@ public List<GroupOfTips>[] GetPrioritizedTipGroups()
 
         public TipInfo GetTipInfo(string globalTipId)
         {
-            return _allTips?[globalTipId];
+            if (!string.IsNullOrEmpty(globalTipId) && _allTips.ContainsKey(globalTipId))
```
Comment (justcla) [inline HotTips/Engine/TipManager.cs]: Is this change necessary? Wouldn't it return null if globalTipId is null or if allTips does not contain the tipId?

File: HotTips/Engine/VSTipHistoryManager.cs [hunk 2]
```diff
@@ -44,12 +46,22 @@ private void InitialiseTipHistoryManager()
...
+                }
+            }
+
+            _tipsSeen = tipsInfo;
 
             SettingsManager.TryGetValue(EXCLUDED_TIP_GROUPS, out string excludedGroups);
 
             _excludedTipGroups = (!string.IsNullOrEmpty(excludedGroups))
-                ? new HashSet<string>(excludedGroups.Split(','), StringComparer.OrdinalIgnoreCase)
+                ? new HashSet<string>(excludedGroups.Split(';'), StringComparer.OrdinalIgnoreCase)
```
Comment (justcla) [inline HotTips/Engine/VSTipHistoryManager.cs]: Not sure you need to modify the way excludedGroups are stored.

File: HotTips/Engine/VSTipHistoryManager.cs [hunk 3]
```diff
@@ -60,27 +72,41 @@ private string GetTipHistoryFromSettingsStore()
             return tipHistoryRaw;
         }
 
-        public void MarkTipAsSeen(string globalTipId)
+        public void MarkTipAsSeen(TipHistoryInfo globalTipId)
```
Comment (justcla) [inline HotTips/Engine/VSTipHistoryManager.cs]: Variable name no longer matches the type.

Comment (justcla) [reply HotTips/Engine/VSTipHistoryManager.cs]: Also, it looks like we only need the globalTipId in this method. Was there a reason to change the input parameter?

File: HotTips/Engine/VSTipHistoryManager.cs [hunk 4]
```diff
@@ -60,27 +72,41 @@ private string GetTipHistoryFromSettingsStore()
...
             // Update the VS settings store with the latest tip history
-            string tipHistoryRaw = String.Join(",", tipsSeen);
+            string tipHistoryRaw = String.Join(";", tipsSeen);
             // Get Writable settings store and update value
             string collectionPath = TIP_OF_THE_DAY_SETTINGS;
             //ISettingsList settingsList = SettingsManager.GetOrCreateList(collectionPath, isMachineLocal: !RoamSettings);
             SettingsManager.SetValueAsync(collectionPath, tipHistoryRaw, isMachineLocal: !RoamSettings);
         }
 
+        public void SaveTipStatus(TipHistoryInfo globalTipId)
```
Comment (justcla) [inline HotTips/Engine/VSTipHistoryManager.cs]: Does this method actually "SaveTipStatus"? It looks like it only adds it to the in-memory object, but does not persist it to the registry.

File: HotTips/Engine/VSTipHistoryManager.cs [hunk 5]
```diff
@@ -115,7 +140,7 @@ public void MarkTipGroupAsExcluded(string tipGroupId)
 
         private void StoreExcludedGroupsToSettings()
         {
-            string value = string.Join(",", _excludedTipGroups);
+            string value = string.Join(";", _excludedTipGroups);
```
Comment (justcla) [inline HotTips/Engine/VSTipHistoryManager.cs]: We didn't really need this change. But it might be OK for consistency.

File: HotTips/TipObjects/TipObjects.cs [hunk 6]
```diff
@@ -68,6 +68,31 @@ public static string GetTipId(string globalTipId)
...
+            return string.Concat(globalTipId, ':', (int)tipLikeStatus);
+        }
+
+        public TipHistoryInfo()
+        {
+
+        }
+        public TipHistoryInfo(string tipHistoryItem)
+        {
+            if (tipHistoryItem.Contains(":"))
```
Comment (justcla) [inline HotTips/TipObjects/TipObjects.cs]: What if this condition is false?

File: HotTips/TipObjects/TipObjects.cs [hunk 7]
```diff
@@ -68,6 +68,31 @@ public static string GetTipId(string globalTipId)
...
+        public override string ToString()
+        {
+            return string.Concat(globalTipId, ':', (int)tipLikeStatus);
+        }
+
+        public TipHistoryInfo()
+        {
+
+        }
+        public TipHistoryInfo(string tipHistoryItem)
```
Comment (justcla) [inline HotTips/TipObjects/TipObjects.cs]: Might be best to take two parameters: globalTipId, likeStatus.

Could have another public static method for "ParseTipHistoryItem" that will return a new TipHistoryInfo or raise exception if unparsable.

File: HotTips/TipObjects/TipObjects.cs [hunk 8]
```diff
@@ -68,6 +68,31 @@ public static string GetTipId(string globalTipId)
         }
     }
 
+    public class TipHistoryInfo
+    {
+        public string globalTipId { get; set; }
+
+        public TipLikeEnum tipLikeStatus { get; set; }
+
+        public override string ToString()
```
Comment (justcla) [inline HotTips/TipObjects/TipObjects.cs]: This is clever, but perhaps not the most correct use ToString. Might be better if this was named differently like "Serialize"

File: HotTips/TipOfTheDay.cs [hunk 9]
```diff
@@ -51,8 +52,8 @@ public static void ShowWindow()
                 // Now show the dialog
                 tipOfTheDayWindow.Show();
 
-                // Mark tip as seen
-                _tipHistoryManager.MarkTipAsSeen(nextTip.globalTipId);
+                // Mark tip as seen and default like status is normal
+                _tipHistoryManager.MarkTipAsSeen(new TipHistoryInfo(nextTip.globalTipId+":1"));
```
Comment (justcla) [inline HotTips/TipOfTheDay.cs]: Appending ":1" in this parameter is not very clean. Would be better to pass the Enum into the constructor.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 10]
```diff
@@ -143,10 +142,10 @@ private void GoToPrevTip()
             }
 
             // Navigate to the previous tip.
-            bool success = NavigateToTip(previousTip, markAsSeen: false);
+            bool success = NavigateToTip(previousTip, markAsSeen: true);
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: We don't want this change. When navigating to a previous tip, it we don't want to markAsSeen, as it will have already been seen.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 11]
```diff
@@ -166,17 +165,19 @@ private TipInfo GetPreviousTip(List<string> tipHistory, int currentTipHistoryInd
                 // Found a tip! Return it.
                 return previousTip;
             }
-
-            // No Previous tip to show. It's possible the tip previously shown no longer exists.
-            // Look for next previous tip. (drop the index back one position and try again)
-            // repeat - until currentTipHistoryIndex < 1 or previousTip != null
-            return GetPreviousTip(tipHistory, prevTipHistoryIndex);
+            else
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: else is redundant. Code will fall through to this block in the else case.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 12]
```diff
@@ -198,7 +199,11 @@ internal bool NavigateToTip(TipInfo nextTip, bool markAsSeen = true)
                 return false;
             }
 
-            currentTip = nextTip.globalTipId;
+            TipHistoryInfo tipHistoryObj = new TipHistoryInfo();
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: This could use a constructor in the TipHistoryInfo class

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 13]
```diff
@@ -226,6 +231,21 @@ private void ShowTipContent(TipInfo nextTip)
             string tipContentString = ReadStringFromFile(contentFilePath);
             // Update the value in the model (which should be reflected in the view through binding)
             _tipViewModel.TipContent = tipContentString;
+
+            List<TipHistoryInfo> tipHistory = _tipHistoryManager.GetTipHistory();
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: This could be extracted to a separate method for "RenderCurrentTipVotingIcons"

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 14]
```diff
@@ -198,7 +199,11 @@ internal bool NavigateToTip(TipInfo nextTip, bool markAsSeen = true)
...
 
-            currentTip = nextTip.globalTipId;
+            TipHistoryInfo tipHistoryObj = new TipHistoryInfo();
+            tipHistoryObj.globalTipId = nextTip.globalTipId;
+            tipHistoryObj.tipLikeStatus = TipLikeEnum.NORMAL;
+
+            currentTip = tipHistoryObj;
 
             // Render the new tip content in the tip viewer
             ShowTipContent(nextTip);
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: it would be useful if nextTip was a TipHistoryInfo item, because then you would not need to look up the history later for the voting.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 15]
```diff
@@ -226,6 +231,21 @@ private void ShowTipContent(TipInfo nextTip)
             string tipContentString = ReadStringFromFile(contentFilePath);
             // Update the value in the model (which should be reflected in the view through binding)
             _tipViewModel.TipContent = tipContentString;
+
+            List<TipHistoryInfo> tipHistory = _tipHistoryManager.GetTipHistory();
+            TipHistoryInfo historyData = tipHistory.Find(a => a.globalTipId.Equals(nextTip.globalTipId));
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: Instead of looking up the history, this method should take a TipHistoryInfo as a parameter so it already has the voting values.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 16]
```diff
@@ -226,6 +231,21 @@ private void ShowTipContent(TipInfo nextTip)
...
+            if (historyData != null)
+            {
+                if (historyData.tipLikeStatus.Equals(TipLikeEnum.LIKE))
+                {
+                    PopulateLikeFilledImage();
+                }
+
+                if (historyData.tipLikeStatus.Equals(TipLikeEnum.DISLIKE))
+                {
+                    PopulateDislikeFilledImage();
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: This relies on having previously called PopulateDefaults.
It would be better if there were methods for "ShowLikeVote" and "ShowDislikeVote" and "ShowNoVote". So the populate logic is all contained in one place.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 17]
```diff
@@ -282,10 +303,12 @@ private void LikeButton_Click(object sender, RoutedEventArgs e)
                 PopulateLikeFilledImage();
             }
             PopulateDislikeImage();
+            _tipHistoryManager.SaveTipStatus(currentTip);
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: The algorithm is:
1. Update the tipsSeen in memory.
2. Persist to registry.
I can only see you doing #2 here.

File: HotTips/TipOfTheDayWindow.xaml.cs [hunk 18]
```diff
@@ -308,6 +332,7 @@ private void PopulateLikeImage()
 
         private void PopulateDislikeFilledImage()
         {
+            currentTip.tipLikeStatus = TipLikeEnum.DISLIKE;
```
Comment (justcla) [inline HotTips/TipOfTheDayWindow.xaml.cs]: Not sure why this would be done here. Seems like the wrong place to be setting current tip value.

File: HotTips/TipOfTheDayWindow.xaml.cs (hunk 14 above)
Comment (sravanthikoneru) [reply HotTips/TipOfTheDayWindow.xaml.cs]: Here the nextTip is a TipInfo. It contains contentUri but TipHistoryInfo does not. Hence, I left as it is.
